### DISCUSSION TITLE
Uno UI on the server?

### DISCUSSION THREAD

**Discussion by xsteel2003** [unoplatform/uno#4963] [Q&A] [answered]
Thanks very much for uno support for webassembly. I am a UWP developer and I am now anxious to distribute some of my features in a browser.
But webassembly currently has a fatal problem: it starts to run too slowly. The first time you want to download up to 10M of data from the server.

I also noticed a project Ooui: https://github.com/praeclarum/Ooui , which provides another way to run XAML and run Xamarin.Forms on a server. In this way, the speed is very fast, really great! Reference: https://github.com/praeclarum/Ooui/wiki/Xamarin.Forms-on-the-Web .

But I'm a UWP developer and I'm not familiar with Xamarin.Forms. I hope uno can run on the server. Is this feasible?

**Comment by jeromelaban** [comment]
Thanks for your interest! 

The current performance is not to be considered at the moment. The current releases of the mono runtime are using the Interpreted mode, which is about 300x to 400x slower than the upcoming AOT mode. Onc AOT is available, we anticipate a massive useability improvement.

As for running on the server, the inner layers of Uno may be too chatty and synchronous for it to be a valid scenario. The goal of wasm is to make it run on the client, so we're going to have to wait for AOT to make all this better!

**Reply by martin-braun** [reply]
Hi @jeromelaban could you please update on the state? The [Windows calculator sample](https://calculator.platform.uno/) still takes over 5 seconds to load, which is really slow for a small app. Are there any further plans for improvement?

**Reply by jeromelaban** [reply]
@martin-braun are the five seconds you are measuring from a first load, or a subsequent load? In all cases, many improvements are coming with .NET 7, we're already testing those with https://gallery.platform.uno. It will soon come to the other apps we're hosting.

**Reply by martin-braun** [reply]
@jeromelaban any load is affected, and I'm talking about the time until the app can be viewed and interacted with (so the time in which the pre-loader is displayed). Platform macOS, Brave Browser. The gallery is even slower for me. I'm not sure if AOT solves this issue, but websites nowadays often deliver cached HTML through SSR (to get at least the view part done very quickly). I really like what Uno offers, especially since it delivers HTML controls instead of drawing on a canvas, but the loading time of those websites feel dated and will not please any indexing bot with successful SEO. Demos need to be blazing fast to gain interest of developers who consider web-first.

**Reply by jeromelaban** [reply]
We're not validating the Brave browser that much as it's not requested a lot. Could you try with chrome, edge or firefox to see if the behavior is different? If Gallery is even slower for you, this is another topic. Could you provide a video of the loading of the site until you can interact with it?

With regards to SSR and indexing, this is not an area we're currently looking into as we're focusing on Web Apps (and not Web Sites). It does not mean that SSR with Uno will not be possible, but we're not there yet.

**Reply by martin-braun** [reply]
@jeromelaban Please check this, the GIF recording speed is accurate (5 seconds load time without cache):
![2022-05-28 12 23 45 AM](https://user-images.githubusercontent.com/24259245/170796256-8d7a5bc3-2db5-4faf-9dfc-31b4ff0bfddf.gif)

And yes, I see how the focus is on Web Apps and it makes sense. It is nice to see that you say that you are not there *yet*. Better loading times can be beneficial for Web Apps as well, but I can also imagine Uno to be an all-rounder. Thanks for your time and the work you put into this.

**Reply by jeromelaban** [reply]
Thanks for the video. The interesting part here is that the progress bar is present during the startup. It should mostly be absent as the browser cache should be used instead.

Do you use a proxy? Or is there something in your path that could alter the caching directives? I'd be interested in getting a network tab capture, to understand why the cache is not working here. Thanks!

**Reply by martin-braun** [reply]
@jeromelaban There was no browser cache, I was visiting the page the first time on Firefox. However, I also see the progress bar on consecutive refreshes, it's just shorter in that case. But it's really about the first visit that needs to improve in loading times, especially for SEO. I use no proxy, I use LTE mobile network on my country: 

```
   Speedtest by Ookla

     Server: Schlueter Onlinedienste - Ruethen (id = 29806)
        ISP: Deutsche Telekom AG
    Latency:    23.34 ms   (3.95 ms jitter)
   Download:   153.86 Mbps (data used: 258.7 MB)
     Upload:    29.18 Mbps (data used: 43.5 MB)
Packet Loss:     0.0%
 Result URL: https://www.speedtest.net/result/c/68d36a05-c75b-4e5c-8e9c-f1fb12d23152
```

No proxy, no VPN. My local firewall whitelists my browsers, I reach the same speeds in the browser. When I load the page with cache, it will load within 3 seconds, but this is still slow for a doc page that features mostly text and a few controls, but I get that loading speeds is not your main focus, since your focus is on WebApps and not traditional web sites. :)

I think the main problem on the gallery page is that it downloads all sub pages and all content into memory. Is it possible to disable the pre-loader animation and implement lazy loading of entire pages and sections so that the loading time of a website can reach 1 or 2 seconds?

**Reply by jeromelaban** [reply]
At this time, no we are not in the position of being able to lazy fetch content (Wasm is currently very limited at any kind of delay loading). Still, not all pages are loaded in memory at load time for the Gallery app, only when navigated to.

First load is indeed not a concern we're focusing on at this time, as we're expecting that it will take some time like many other Web Apps that need to fetch their large payload and is then stored for subsequent loads.

I understand what your concerns are, and we are not in a position to answer those yet.

**Comment by Flem100** [comment]
I would also like Uno on the server, the same way as Ooui. But sadly I think Ooui is dead, so I might have to put my hopes into Uno :).

**Comment by jeromelaban** [comment]
For now, we have prioritized any work for server-side Uno. What would be your use case?

**Comment by Flem100** [comment]
Hi Jeromelaban. Thank you for your response! :)
The original purpose for Frank when he created Ooui, was to have a simple server side UI, without the need for WebAPIs and and all that serverside vs clientside separation. Just a simple way to display something on the web, like Web Forms (if you don't try to make something too fancy).
Blazor has been released for serverside, but it still uses razor pages and HTML, which is fine, but not exactly what I want. Ooui serverside with all sorts of WPF-controls is exactly what I want :). What do you think? :)

**Comment by francoistanguay** [comment, marked as answer]
XAML is intrinsically synchronous in nature. 
While we appreciate the idea, we have no plans at this point in time to make any server-side solution.
